bin/consequence_mapping/consequence_mapping.py

```diff
@@ -0,0 +1,154 @@
+    """Converts a VEP compatible identifier to colon-based."""
+    vep_id_fields = vep_id.split(' ')
+    return ':'.join([vep_id_fields[0], vep_id_fields[1], vep_id_fields[3], vep_id_fields[4]])
+
+
+@retry(requests.HTTPError, tries=10, delay=5, backoff=1.2, jitter=(1, 3), logger=logger)
+def query_vep(variants, search_distance):
+    """Query VEP and return results in JSON format. Upstream/downstream genes are searched up to a given distance."""
+    ensembl_request_url = 'https://rest.ensembl.org/vep/human/region'
+    headers = {'Content-Type': 'application/json', 'Accept': 'application/json'}
+    result = requests.post(ensembl_request_url, headers=headers, data=json.dumps({
+        'variants': variants, 'distance': search_distance,
+    }))
+    # If there was an HTTP error, raise an exception. This will be caught by @retry
+    result.raise_for_status()
+    return result.json()
+
+
+def extract_consequences(vep_results, acceptable_biotypes, only_closest, results_by_variant):
+    """Given VEP results, return a list of consequences matching certain criteria.
+
+    Args:
+        vep_results: results obtained from VEP for a list of variants, in JSON format.
+        acceptable_biotypes: a list of transcript biotypes to consider (as defined in Ensembl documentation, see
+            https://www.ensembl.org/info/genome/genebuild/biotypes.html). Consequences for other transcript biotypes
+            will be ignored.
+        only_closest: if this flag is specified, then at most one consequence per variant will be returned. The
+            consequences are sorted by distance from the gene and the closest one is chosen. In case of a tie,
+            consequence is selected at random. If this flag is not specified, all consequences for each variant will
+            be returned.
+        results_by_variant: a dict to write the results into.
+    """
+    for result in vep_results:
+        variant_identifier = result['input']
+        results_by_variant.setdefault(variant_identifier, [])
+        consequences = result.get('transcript_consequences', [])
+
+        # Keep only consequences with the allowed biotypes; if there are none, skip this variant
+        consequences = [c for c in consequences if c['biotype'] in acceptable_biotypes]
+        if not consequences:
+            continue
+
+        # Flatten the list of consequence terms and find the most severe one
+        all_consequence_terms = [term for c in consequences for term in c['consequence_terms']]
+        all_consequence_terms.sort(key=lambda term: consequence_term_severity_rank[term])
+        most_severe_consequence_term = all_consequence_terms[0]
+
+        # Keep only consequences which include the most severe consequence term; sort by increasing order of distance
+        consequences = [c for c in consequences if most_severe_consequence_term in c['consequence_terms']]
+        consequences.sort(key=lambda consequence: abs(consequence.get('distance', 0)))
```

Comment:
so we are making the least possible distance as the default? Shouldn't we do it the other way round?

Reply:
Ah, the reason for that is that if there is no `distance` attribute in VEP results, it means that it is not applicable to the variant because it resides *inside* the gene. Hence, in this case the distance is set to 0. I added a comment with this clarification to the code.